### ISSUE TITLE
Testing overloads

### ISSUE THREAD

**Issue by wellwelwel** [tstyche/tstyche#264]
Hi @tstyche, thanks for this project 🙋🏻‍♂️

I'd like to know if it's possible to test overloads. I tried to find it in the docs, but I couldn't find anything related to the word "overload".

Maybe run the function that causes each overload to return in a variable and then test that variable's type, since `ReturnType` doesn't handle overloads.

**Comment by mrazauskas** [comment]
Yes, all TSTyche assertions do support overloads. That is not mentioned explicitly, because TSTyche supports all TypeScript features.

---

What are you trying to test?

An overloaded function type can be declared and test like this:

```ts
import { expect } from "tstyche";

declare function test(cb: () => Promise<unknown>): Promise<void>;
declare function test(cb: () => unknown): void;

expect(test).type.toBe<{
  (cb: () => Promise<unknown>): Promise<void>;
  (cb: () => unknown): void;
}>();
```

The declaration syntax can be different, but the resulting type is the same. Here I try to show the syntax of a type which represents an overloaded function. Note that this test looks somewhat like a snapshot. The shape of the type gets checked, but it does not proof that the type works as expected.

As far as I know, TypeScript does not have a type that represents return type of an overloaded function. One of the goals of TSTyche is to stay idiomatic, hence it would be odd to introduce syntax that does not exist in TypeScript.

Also just like checking the whole shape of overloaded function, checking its return type does not proof that the type works as expect. Just like with functional tests, it is the best to look at things from the users perspective. So:

```ts
import { expect } from "tstyche";

declare function test(cb: () => Promise<unknown>): Promise<void>;
declare function test(cb: () => unknown): void;

expect(test(() => {})).type.toBe<void>();
expect(test(() => Promise.resolve())).type.toBe<Promise<void>>();
```

---

Thanks for the question. I plan to add more guides to the docs at some point. For now I will move this to discussions.

Let me know, if you had something else in mind. Always happy to think about type tests (;
